compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java

```diff
@@ -178,6 +194,54 @@ public ComplexMatchResult multiplyAddSub(BinaryNode binary, ValueNode a, ValueNo
         return builder -> getArithmeticLIRGenerator().emitMSub(operand(a), operand(b), operand(c));
     }
 
+    /**
+     * If x < 0:
+     *
+     * -> -((-x) & 1) == 0.
+     *
+     * else:
+     *
+     * -> ((((x >> (bitCount - 1)) >>> (bitCount - 1)) + x) & (-2)) == x.
+     */
+    @MatchRule("(If (IntegerEquals (Negate=cmpLeft (And (Negate value) Constant=a)) Constant=b))")
+    @MatchRule("(If (IntegerEquals (And=cmpLeft (Add (UnsignedRightShift (RightShift value Constant=a) Constant=a) value) Constant=b) value))")
+    public ComplexMatchResult integerEqualsOddEvenBranch(IfNode root, ValueNode cmpLeft, ValueNode value, ConstantNode a, ConstantNode b) {
+        if (!value.getStackKind().isNumericInteger()) {
+            return null;
+        }
+
+        long constA = a.asJavaConstant().asLong();
+        long constB = b.asJavaConstant().asLong();
+
+        if (cmpLeft instanceof NegateNode) {
+            IntegerStamp stamp = (IntegerStamp) value.stamp(NodeView.DEFAULT);
+            if (stamp.isNegative() && constA == 1 && constB == 0) {
+                return emitBitTestAndBranch(root, value);
+            }
+        }
+
+        if (cmpLeft instanceof AndNode) {
+            if (constA == value.getStackKind().getBitCount() - 1 && constB == -2) {
+                return emitBitTestAndBranch(root, value);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * x % 2 == 0 -> x & 1 == 0.
+     */
+    @MatchRule("(If (IntegerTest value Constant=a))")
+    public ComplexMatchResult integerTestOddEvenBranch(IfNode root, ValueNode value, ConstantNode a) {
+        if (value.getStackKind().isNumericInteger()) {
```

Comment:
I think we could extend this pattern to any bit set - this is what `tbz` and `tbnz` actually do. I think of:
```
long constant = a.asJavaConstant();
if(Long.bitCount(constant) == 1) {
  int bitToTest = Long.numberOfTrailingZeros(constant);
  emitBitTestAndBranch(root, value, bitToTest);
}
```

Reply:
Yes, make sense!

Reply:
Hi sanzinger, I extend this rule as you suggested. But I tested the  `AArch64TestBitAndBranchTest.testBit42` and it failed. The `bitToTest = Long.numberOfTrailingZeros(constant)` is 41, but the finally assembly code is `tbz #5`. It seems that there is something wrong with the tbz assembler function. I tried `(a & (1 << 7)) == 0` and the result is `tbz #3`.